In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
df = pd.read_csv('./data/netflix_user_data_unclean.csv')

In [3]:
df.describe()

Age  Unnamed: 16   Unnamed: 17  Unnamed: 18
count  118542.000000          0.0      2.000000          1.0
mean       43.823716          NaN  21954.000000       4010.0
std        15.317243          NaN   2835.498193          NaN
min        18.000000          NaN  19949.000000       4010.0
25%        30.000000          NaN  20951.500000       4010.0
50%        44.000000          NaN  21954.000000       4010.0
75%        57.000000          NaN  22956.500000       4010.0
max        70.000000          NaN  23959.000000       4010.0

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 119779 entries, 0 to 119778
Data columns (total 19 columns):
 #   Column                              Non-Null Count   Dtype  
---  ------                              --------------   -----  
 0   Customer ID                         119378 non-null  object 
 1   Subscription Length (Months)        119379 non-null  object 
 2   Customer Satisfaction Score (1-10)  119369 non-null  object 
 3   Daily Watch Time (Hours)            119352 non-null  object 
 4   Engagement Rate (1-10)              119354 non-null  object 
 5   Device Used Most Often              119332 non-null  object 
 6   Genre Preference                    119342 non-null  object 
 7   Region                              119380 non-null  object 
 8   Payment History (On-Time/Delayed)   119330 non-null  object 
 9   Subscription Plan                   119383 non-null  object 
 10  Churn status                        119391 non-null  object 
 11  Support Queries Logged    

In [5]:
df.isnull().sum()

Customer ID                              401
Subscription Length (Months)             400
Customer Satisfaction Score (1-10)       410
Daily Watch Time (Hours)                 427
Engagement Rate (1-10)                   425
Device Used Most Often                   447
Genre Preference                         437
Region                                   399
Payment History (On-Time/Delayed)        449
Subscription Plan                        396
Churn status                             388
Support Queries Logged                   430
Age                                     1237
Monthly Income ($)                       419
Promotional Offers Used                  443
Number of Profiles Created               398
Unnamed: 16                           119779
Unnamed: 17                           119777
Unnamed: 18                           119778
dtype: int64

In [6]:
print(df[['Unnamed: 16', 'Unnamed: 17', 'Unnamed: 18']])

        Unnamed: 16  Unnamed: 17  Unnamed: 18
0               NaN          NaN          NaN
1               NaN          NaN          NaN
2               NaN          NaN          NaN
3               NaN          NaN          NaN
4               NaN          NaN          NaN
...             ...          ...          ...
119774          NaN          NaN          NaN
119775          NaN          NaN          NaN
119776          NaN          NaN          NaN
119777          NaN          NaN          NaN
119778          NaN          NaN          NaN

[119779 rows x 3 columns]


In [7]:
df.drop(columns=['Unnamed: 16', 'Unnamed: 17', 'Unnamed: 18'], inplace=True) # 불필요한 열 제거(전부 nan값)

In [8]:
df.describe() # 수치형 데이터 요약 통계량 확인

Age
count  118542.000000
mean       43.823716
std        15.317243
min        18.000000
25%        30.000000
50%        44.000000
75%        57.000000
max        70.000000

오브젝트 값 확인하고 숫자형, 문자형 구분

 0   Customer ID                         119378 non-null  object 

 1   Subscription Length (Months)        119379 non-null  object 

 2   Customer Satisfaction Score (1-10)  119369 non-null  object 

 3   Daily Watch Time (Hours)            119352 non-null  object 

 4   Engagement Rate (1-10)              119354 non-null  object 

 5   Device Used Most Often              119332 non-null  object 
 
 6   Genre Preference                    119342 non-null  object 

 7   Region                              119380 non-null  object 

 8   Payment History (On-Time/Delayed)   119330 non-null  object 

 9   Subscription Plan                   119383 non-null  object 

 10  Churn status                        119391 non-null  object 

 11  Support Queries Logged              119349 non-null  object 

 13  Monthly Income ($)                  119360 non-null  object 

 14  Promotional Offers Used             119336 non-null  object 

 15  Number of Profiles Created          119381 non-null  object 

---
숫자열 컬럼

In [9]:
print(df['Customer ID'].value_counts()) # 고객 ID의 고유값 개수 확인(중복된 고객이 있는지 확인)

Customer ID
###        413
C028760      2
C013775      2
C024383      2
C030099      2
          ... 
C118031      1
C118032      1
C118033      1
C118034      1
C000006      1
Name: count, Length: 117245, dtype: int64


In [10]:
# 결측치 401개 존재
print(len(df)) # 전체 행 개수 확인
print(df['Customer ID'].nunique()) # 고객 ID의 고유값 개수 확인(중복된 고객이 있는지 확인)
df = df.drop_duplicates(subset=['Customer ID']) # 고객 ID를 기준으로 중복된 행 제거
print(len(df)) # 중복 제거 후 행 개수 확인

119779
117245
117246


In [11]:
print(df['Subscription Length (Months)'].value_counts()) # 구독 기간 분포 확인

Subscription Length (Months)
6      25988
3      23070
24     23000
12     22289
1      22003
###      367
2z        20
x4        18
2x        18
2y        17
x2        15
1x        14
1y        13
z4        11
y2        10
1z         9
z2         7
y4         7
Name: count, dtype: int64


In [12]:
df['Subscription Length (Months)'] = pd.to_numeric(df['Subscription Length (Months)'], errors='coerce') # 숫자형으로 변환, 변환 불가 시 NaN 처리(컬럼명을 월 단위로 명시되어 있기 때문에 나머지 값 제거)

In [13]:
print(df['Customer Satisfaction Score (1-10)'].value_counts()) # 고객 만족도 점수 분포 확인

Customer Satisfaction Score (1-10)
2     13291
4     12471
5     12463
3     12130
6     11975
10    11628
8     11065
1     11061
9     10372
7     10013
15      352
1z       12
x0       10
1x        9
z0        7
y0        4
1y        3
Name: count, dtype: int64


In [14]:
df['Customer Satisfaction Score (1-10)'] = pd.to_numeric(df['Customer Satisfaction Score (1-10)'], errors='coerce') # 숫자형으로 변환, 변환 불가 시 NaN 처리(컬럼 명이 1~10으로 명시되어 있기 때문에 나머지 값 제거)

In [15]:
df = df[(df['Customer Satisfaction Score (1-10)'] >= 1) & (df['Customer Satisfaction Score (1-10)'] <= 10)] # 고객 만족도 점수가 1~10 사이인 행만 남김

In [16]:
print(df['Daily Watch Time (Hours)'].value_counts()) # 일일 시청 시간 분포 확인

Daily Watch Time (Hours)
4.59    1037
3.25     815
0.87     804
0.77     699
3.11     697
        ... 
3.z3       1
x.26       1
2y99       1
4z58       1
3z26       1
Name: count, Length: 721, dtype: int64


In [17]:
df['Daily Watch Time (Hours)'] = pd.to_numeric(df['Daily Watch Time (Hours)'], errors='coerce') # 숫자형으로 변환, 변환 불가 시 NaN 처리(컬럼 명이 시간 단위로 명시되어 있기 때문에 나머지 값 제거)

In [18]:
print(df['Engagement Rate (1-10)'].value_counts()) # 참여도 점수 분포 확인

Engagement Rate (1-10)
9     13413
6     13198
2     13076
8     11223
5     11213
3     11125
1     10876
10    10636
4     10520
7     10418
-2      364
1x       11
y0        6
z0        5
1z        5
1y        4
x0        1
Name: count, dtype: int64


In [19]:
df['Engagement Rate (1-10)'] = pd.to_numeric(df['Engagement Rate (1-10)'], errors='coerce') # 숫자형으로 변환, 변환 불가 시 NaN 처리(컬럼 명이 1~10으로 명시되어 있기 때문에 나머지 값 제거)

In [20]:
df.loc[df['Engagement Rate (1-10)'] < 0, 'Engagement Rate (1-10)'] = np.nan # 참여도 점수가 0 미만인 경우 NaN 처리

In [21]:
df = df[(df['Engagement Rate (1-10)'] >= 1) & (df['Engagement Rate (1-10)'] <= 10)] # 고객 만족도 점수가 1~10 사이인 행만 남김

In [22]:
print(df['Support Queries Logged'].value_counts()) # 지원 문의 기록 분포 확인

Support Queries Logged
3       11750
6       11728
4       10925
7       10805
9       10589
0       10588
5       10333
2       10206
10       9636
1        9561
8        8845
Many      349
1y          8
1z          6
x0          6
z0          5
1x          5
y0          1
Name: count, dtype: int64


In [23]:
print(df['Support Queries Logged'].value_counts()) # 지원 문의 기록 분포 확인

Support Queries Logged
3       11750
6       11728
4       10925
7       10805
9       10589
0       10588
5       10333
2       10206
10       9636
1        9561
8        8845
Many      349
1y          8
1z          6
x0          6
z0          5
1x          5
y0          1
Name: count, dtype: int64


In [24]:
df['Support Queries Logged'] = pd.to_numeric(df['Support Queries Logged'], errors='coerce') # 숫자형으로 변환, 변환 불가 시 NaN 처리(many 값은 표현상 10이상인 것 같지만 표본이 적어 제거)

In [25]:
print(df['Monthly Income ($)'].value_counts()) # 월 소득 분포 확인

Monthly Income ($)
3085       462
6014       345
4086       345
Unknown    321
3942       234
          ... 
y120         1
55z6         1
9y48         1
1y30         1
x611         1
Name: count, Length: 1209, dtype: int64


In [26]:
numerlic_monthly_income = df['Monthly Income ($)'].astype(str).str.contains(r'[^0-9.]', na=False) # 숫자와 소수점 이외의 문자가 포함된 행 찾기
df = df[~numerlic_monthly_income] # 숫자와 소수점 이외의 문자가 포함된 행 제거
df['Monthly Income ($)'] = pd.to_numeric(df['Monthly Income ($)'], errors='coerce') # 숫자형으로 변환, 변환 불가 시 NaN 처리
print(df['Monthly Income ($)'].value_counts()) # 월 소득 분포 확인

Monthly Income ($)
3085    462
6014    345
4086    345
3942    234
5577    234
       ... 
2426    109
6736    109
747     109
4952    109
5872    109
Name: count, Length: 933, dtype: int64


In [27]:
print(df['Promotional Offers Used'].value_counts()) # 사용한 프로모션 오퍼 분포 확인

Promotional Offers Used
0      19964
1      19661
3      19514
2      19315
4      18932
5      16771
###      286
Name: count, dtype: int64


In [28]:
df['Promotional Offers Used'] = pd.to_numeric(df['Promotional Offers Used'], errors='coerce') # 숫자형으로 변환, 변환 불가 시 NaN 처리

In [29]:
print(df['Number of Profiles Created'].value_counts()) # 생성한 프로필 수 분포 확인

Number of Profiles Created
1      25541
4      23659
2      22367
5      21408
3      21214
###      287
Name: count, dtype: int64


In [30]:
df['Number of Profiles Created'] = pd.to_numeric(df['Number of Profiles Created'], errors='coerce') # 숫자형으로 변환, 변환 불가 시 NaN 처리

In [31]:
df.describe()

Subscription Length (Months)  Customer Satisfaction Score (1-10)  \
count                 114059.000000                       114764.000000   
mean                       9.169491                            5.378385   
std                        8.212657                            2.854364   
min                        1.000000                            1.000000   
25%                        3.000000                            3.000000   
50%                        6.000000                            5.000000   
75%                       12.000000                            8.000000   
max                       24.000000                           10.000000   

       Daily Watch Time (Hours)  Engagement Rate (1-10)  \
count             113800.000000           114764.000000   
mean                   2.828675                5.510909   
std                    1.302063                2.870003   
min                    0.500000                1.000000   
25%                    1.650000                3.000000   
50%                    2.890000                6.000000   
75%                    3.930000                8.000000   
max                    5.000000               10.000000   

       Support Queries Logged            Age  Monthly Income ($)  \
count           114095.000000  113845.000000       114764.000000   
mean                 4.949314      43.846335         5251.483035   
std                  3.110216      15.346386         2719.714973   
min                  0.000000      18.000000          510.000000   
25%                  2.000000      30.000000         3071.000000   
50%                  5.000000      44.000000         5272.000000   
75%                  8.000000      57.000000         7596.000000   
max                 10.000000      70.000000         9994.000000   

       Promotional Offers Used  Number of Profiles Created  
count            114157.000000               114189.000000  
mean                  2.421367                    2.938926  
std                   1.688414                    1.429672  
min                   0.000000                    1.000000  
25%                   1.000000                    2.000000  
50%                   2.000000                    3.000000  
75%                   4.000000                    4.000000  
max                   5.000000                    5.000000

---
문자열

In [32]:
print(df['Device Used Most Often'].value_counts()) # 가장 자주 사용하는 기기 분포 확인

Device Used Most Often
Laptop      25597
Mobile      22871
Tablet      22751
Smart TV    21520
Desktop     21078
            ...  
Desktox         1
Tabyet          1
Tablex          1
Deskyop         1
Mobilz          1
Name: count, Length: 97, dtype: int64


In [33]:
valid_devices = ['Laptop', 'Mobile', 'Tablet', 'Smart TV', 'Desktop'] # 유효한 기기 목록
df = df[df['Device Used Most Often'].isin(valid_devices)] # 유효한 기기만 남기고 나머지 제거
print(df['Device Used Most Often'].value_counts()) # 가장 자주 사용하는 기기 분포 확인

Device Used Most Often
Laptop      25597
Mobile      22871
Tablet      22751
Smart TV    21520
Desktop     21078
Name: count, dtype: int64


In [34]:
print(df['Genre Preference'].value_counts()) # 선호 장르 분포 확인

Genre Preference
Sci-Fi         18870
Romance        16318
Drama          16275
Thriller       15836
Documentary    15671
               ...  
Thyiller           1
Documentxry        1
ScixFi             1
Thrillex           1
Docxmentary        1
Name: count, Length: 126, dtype: int64


In [35]:
valid_genres = ['Sci-Fi', 'Romance', 'Drama', 'Thriller', 'Documentary', 'Comedy'] # 유효한 장르 목록
df = df[df['Genre Preference'].isin(valid_genres)] # 유효한 장르만 남기고 나머지 제거
print(df['Genre Preference'].value_counts()) # 선호 장르 분포 확인

Genre Preference
Sci-Fi         18870
Romance        16318
Drama          16275
Thriller       15836
Documentary    15671
Comedy         14691
Name: count, dtype: int64


In [36]:
pd.set_option('display.max_rows', None) # 모든 행 출력 설정
print(df['Region'].value_counts()) # 지역 분포 확인
pd.reset_option('display.max_rows') # 설정 초기화

Region
Asia             20417
South America    20100
Europe           19563
Africa           18851
North America    18090
Eurasia            199
Asiy                 7
Afyica               7
Ezrope               6
Afzica               5
Afriza               5
Azia                 5
Eurype               5
Ayia                 5
South Amyrica        5
Euyope               5
Asix                 5
Asya                 4
Africy               4
zurope               4
South Ayerica        4
Europx               4
Afxica               4
Soxth America        4
yfrica               4
Soutz America        3
Axia                 3
North Axerica        3
South Ameriya        3
Afriya               3
Azrica               3
zfrica               3
Ayrica               3
North Americy        3
yurope               3
Souyh America        3
North Amexica        3
xurope               3
zsia                 3
ysia                 3
xsia                 3
Exrope               3
Asxa                 3
Asiz

In [37]:
valid_regions = ['Asia', 'Africa', 'Europe', 'South America', 'North America', 'Eurasia'] # 유효한 지역 값 리스트
df = df[df['Region'].isin(valid_regions)] # 유효한 지역 값만 남기고 나머지 행 제거
print(df['Region'].value_counts()) # 지역 분포 확인

Region
Asia             20417
South America    20100
Europe           19563
Africa           18851
North America    18090
Eurasia            199
Name: count, dtype: int64


In [38]:
print(df['Payment History (On-Time/Delayed)'].value_counts()) # 결제 이력 분포 확인
df['Payment History (On-Time/Delayed)'].replace('Late', 'Delayed', inplace=True) # 'Late' 값을 'Delayed'로 통일
valid_payment_history = ['On-Time', 'Delayed'] # 유효한 결제 이력 값 리스트
df = df[df['Payment History (On-Time/Delayed)'].isin(valid_payment_history)] # 유효한 결제 이력 값만 남기고 나머지 행 제거

Payment History (On-Time/Delayed)
Delayed    49605
On-Time    47006
Late         210
On-Tiye       11
Delayex       10
On-Timx        9
On-Tzme        8
On-Tyme        8
Delzyed        7
Delaxed        6
On-Tixe        6
Deyayed        6
Delazed        6
Dylayed        6
Delayyd        5
Delayez        5
yelayed        5
xn-Time        4
OnzTime        4
Oy-Time        4
zelayed        4
zn-Time        4
yn-Time        4
Delayxd        4
Dxlayed        4
Delayzd        4
OnyTime        4
Delyyed        4
On-Tize        3
OnxTime        3
Dezayed        3
xelayed        3
Delayey        3
Oz-Time        3
On-Timy        3
Delxyed        3
Dexayed        3
On-zime        3
On-Txme        2
On-Timz        2
Ox-Time        2
On-yime        1
Dzlayed        1
Name: count, dtype: int64


C:\Users\Playdata\AppData\Local\Temp\ipykernel_22064\1789514189.py:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['Payment History (On-Time/Delayed)'].replace('Late', 'Delayed', inplace=True) # 'Late' 값을 'Delayed'로 통일


In [39]:
print(df['Payment History (On-Time/Delayed)'].value_counts()) # 결제 이력 분포 확인

Payment History (On-Time/Delayed)
Delayed    49815
On-Time    47006
Name: count, dtype: int64


In [40]:
print(df['Subscription Plan'].value_counts()) # 구독 플랜 분포 확인

Subscription Plan
Standard    33296
Premium     32746
Basic       30225
Standrd       174
Staxdard        8
            ...  
ytandard        1
Sxandard        1
Premizm         1
xtandard        1
Standarx        1
Name: count, Length: 61, dtype: int64


In [41]:
valid_plans = ['Basic', 'Standard', 'Premium'] # 유효한 구독 플랜 값 리스트
df = df[df['Subscription Plan'].isin(valid_plans)] # 유효한 구독 플랜 값만 남기고 나머지 행 제거
print(df['Subscription Plan'].value_counts()) # 구독 플랜 분포 확인

Subscription Plan
Standard    33296
Premium     32746
Basic       30225
Name: count, dtype: int64


In [42]:
print(df['Churn status'].value_counts()) # 이탈 상태 분포 확인

Churn status
Yes      86172
No        9636
Maybe      166
zo          15
Yey         13
Yex         12
Ny          12
yo          12
Nx          10
Yxs         10
xes          9
Nz           9
xo           8
zes          7
Yez          7
yes          6
Yzs          6
Yys          3
Name: count, dtype: int64


In [43]:
valid_Churn_status = ['Yes', 'No'] # 유효한 이탈 상태 값 리스트
df = df[df['Churn status'].isin(valid_Churn_status)] # 유효한 이탈 상태 값만 남기고 나머지 행 제거
print(df['Churn status'].value_counts()) # 이탈 상태 분포 확인

Churn status
Yes    86172
No      9636
Name: count, dtype: int64


In [44]:
print(df.isnull().sum()) # 각 열의 결측치 개수 확인

Customer ID                             0
Subscription Length (Months)          333
Customer Satisfaction Score (1-10)      0
Daily Watch Time (Hours)              459
Engagement Rate (1-10)                  0
Device Used Most Often                  0
Genre Preference                        0
Region                                  0
Payment History (On-Time/Delayed)       0
Subscription Plan                       0
Churn status                            0
Support Queries Logged                313
Age                                   455
Monthly Income ($)                      0
Promotional Offers Used               293
Number of Profiles Created            270
dtype: int64


In [45]:
df.dropna(inplace=True) # 결측치가 있는 행 제거

In [46]:
print(df.isnull().sum()) # 각 열의 결측치 개수 확인

Customer ID                           0
Subscription Length (Months)          0
Customer Satisfaction Score (1-10)    0
Daily Watch Time (Hours)              0
Engagement Rate (1-10)                0
Device Used Most Often                0
Genre Preference                      0
Region                                0
Payment History (On-Time/Delayed)     0
Subscription Plan                     0
Churn status                          0
Support Queries Logged                0
Age                                   0
Monthly Income ($)                    0
Promotional Offers Used               0
Number of Profiles Created            0
dtype: int64


In [47]:
cleaned_file_path = './data/netflix_user_data_cleaned.csv'
df.to_csv(cleaned_file_path, index=False) # 정제된 데이터셋을 새로운 CSV 파일로 저장